### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by amjltc295** [IBM/pytorch-seq2seq#183]
I notice that there seems to be a memory leak issue during training. 

On my K80, I set the batch size to 24 and the GPU memory consumption is about 4000MB. However, as the training process goes on,  the GPU memory consumptions increases and a `RuntimeError: CUDA error: out of memory` is raised about 30 minutes later. If I set the batch size to 16 the error would not take place but the memory still increase.

**Comment by qiandai798** [comment]
> I notice that there seems to be a memory leak issue during training.
> 
> On my K80, I set the batch size to 24 and the GPU memory consumption is about 4000MB. However, as the training process goes on, the GPU memory consumptions increases and a `RuntimeError: CUDA error: out of memory` is raised about 30 minutes later. If I set the batch size to 16 the error would not take place but the memory still increase.

I've faced with the same problem. My modified networks is about 600MB and during training my GPU memory consumption become stable at around 26500MB. So when I was using 1080ti, I ran out of memory before the consumption became stable in less than an epoch. I can run it on GV100 now. I'm not sure but it seems to have something to do with the increasing of sentence lengths if you sort all training data by lengths of sentences. But 16 batch_size seems to be a little too small.

**Comment by YupengGao** [comment]
@amjltc295 I have the same issue RuntimeError: CUDA error: out of memory. Have you solve the problem yet?
or do you have any idea how to debug this kind of error?

**Comment by amjltc295** [comment]
@YupengGao  No, I just keep the batch size small. I wish someone could give me a clue.
For debugging, https://discuss.pytorch.org/t/how-to-debug-causes-of-gpu-memory-leaks/6741/3 might help
